core/src/main/scala/cats/SelectiveError.scala

```diff
@@ -0,0 +1,15 @@
+package cats
+
+import cats.syntax.either._
+
+trait SelectiveError[F[_], E] extends ApplicativeError[F, E] with Selective[F] {
+
+  /**
+   * Turns a successful value into an error if it does not satisfy a given predicate.
+   */
+  def ensure[A](fa: F[A])(error: => E)(predicate: A => Boolean): F[A] =
```

Comment:
Maybe comment why error is lazy but we always evaluate it by calling raiseError

Reply:
I overlooked the laziness.  And that laziness is useful if `error` is `Throwable`, so we don't pay for stack traces on the happy path.  Maybe this is just a new function without laziness?

I don't yet know whether `SelectiveError` is a useful abstraction.  This was a "because we can."

Reply:
I would be conservative and not do "because we can".

Reply:
I prefer not to lean too hard on Haskell, but there's no precedent there to justify its usefulness.  I'll remove it for now.